Fix all instances of -Waddress-of-packed-member

carlescufi opened the pull request:
This finally closes #16587 by removing all remaining pieces of code that can trigger the compiler warning after #39192 resolved all the ones in the networking code.

Additional analysis of the problem can be found [here](https://github.com/zephyrproject-rtos/zephyr/issues/16587#issuecomment-930116330).

carlescufi committed:
gcc: Enable address-of-packed-member warnings

After having resolved all of the instances of packed member access,
re-enable the warning.

Fixes #16587.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
kernel: mmu: Fix access to unpacked member inside packed struct

The following warning is triggered by GCC when
-Waddress-of-packed-member is enabled:

/home/carles/src/zephyr/zephyr/kernel/mmu.c: In function
'free_page_frame_list_put':
/home/carles/src/zephyr/zephyr/kernel/mmu.c:383:42: warning: taking
address of packed member of 'struct z_page_frame' may result in an
unaligned pointer value [-Waddress-of-packed-member]
  383 |  sys_slist_append(&free_page_frame_list, &pf->node);

This is due to the fact that sys_snode_t node is an unpacked structure
inside a packed z_page_frame structure, so that the alignment of the
former cannot be ensured if placed inside the latter.

Given that alignment of z_page_frame is ensured by the code, silence the
compiler by going through an intermediate variable.

More info in #16587.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
include: Add a macro to check pointer alignment

Add a new macro, IS_PTR_ALIGNED() that verifies if a pointer is aligned
sufficiently for a particular data type provided as an argument.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
Bluetooth: controller: Fix access to members of packed structs

Fix all instances in the controller of -Waddress-of-packed-member by
casting through an intermediate variable and verifying the alignment
with an assertion.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>
Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

carlescufi commented on include/toolchain/common.h:
```diff
@@ -187,6 +187,9 @@
 #define Z_DECL_ALIGN(type) __aligned(__alignof(type)) type
 
+/* Check if a pointer is aligned enough for a particular data type. */
+#define IS_PTR_ALIGNED(ptr, type) ((((uintptr_t)ptr) % __alignof(type)) == 0)
```
@rlubos it would be possible to replace some of the `UNALIGNED_GET()` code in networking with usage of this macro inside an assert statement.

rlubos replied:
Do you have any examples where could this be applied? I had an impression, that with data inside a packet (which the packed struct is casted on) we should not make any assumptions regarding alignment?

carlescufi replied:
No, only in places where you know or suspect that the alignment has been ensured by the allocation itself. If you do not know where the allocation was performed then it wouldn't be useful. But I thought we had some of those in networking, maybe I was wrong.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -2868,8 +2869,11 @@ static inline void event_conn_upd_init(struct ll_conn *conn,
 
-		conn->llcp.conn_upd.pdu_win_offset = (uint16_t *)
-			&pdu_ctrl_tx->llctrl.conn_update_ind.win_offset;
+		win_offs = (void *)
+				&pdu_ctrl_tx->llctrl.conn_update_ind.win_offset;
+
+		LL_ASSERT(IS_PTR_ALIGNED(win_offs, uint16_t));
+		conn->llcp.conn_upd.pdu_win_offset = (void *)win_offs;
```
This is intentionally misaligned as the field is inside a packed PDU struct. memcpy is used to fill this address ensuring there is no memfault.

carlescufi replied:
I assume this is true for line 3759, which is currently asserting?

cvinayak replied:
3759 in this PR?  None of the struct typecast use should assert, if they do, then its genuine misalignment/bug.

carlescufi replied:
The assert in line 3759 in ull_conn.c hits, but it's not a struct, it's an int16_t @cvinayak \

cvinayak replied:
> I assume this is true too for line 3759, which is currently asserting?

Yes.

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -6628,7 +6628,11 @@ static void le_scan_req_received(struct pdu_data *pdu_data,
 			     struct net_buf *buf)
 {
-	struct node_rx_cc *cc = (void *)pdu_data;
+	void *node = (void *)pdu_data;
+
+	LL_ASSERT(IS_PTR_ALIGNED(node, struct node_rx_cc));
+
+	struct node_rx_cc *cc = node;
```
This: https://github.com/zephyrproject-rtos/zephyr/blob/aa83b8d6e3bb5e9fbe35912350e9faa6c6e9c5b2/subsys/bluetooth/controller/ll_sw/lll.h#L391-L401

should have avoided misaligned access?

carlescufi replied:
No, because `node_rx_cc` and `node_rx_pdu` are not packed and have alignment requirements, whereas `struct pdu_data` is packed and thus has zero alignment requirements. Hence, the compiler complains.

cvinayak replied:
`__aligned(4)` ensures that inside `strict node_rx_pdu` the field `pdu` is  not unaligned. The allocation by `mem_acquire` ensures runtime aligned memory for each node.

>  because node_rx_cc and node_rx_pdu are not packed and have alignment requirements, whereas struct pdu_data is packed and thus has zero alignment requirements. Hence, the compiler complains.

Agree; on the compile time complain.

carlescufi committed:
kernel: userspace: Fix address-of-packed-mem warning

The warning below appears once -Waddress-of-packed-mem is enabled:

/home/carles/src/zephyr/zephyr/kernel/userspace.c: In function
'unref_check':
/home/carles/src/zephyr/zephyr/kernel/userspace.c:471:28: warning:
converting a packed 'struct z_object' pointer (alignment 4) to a 'struct
dyn_obj' pointer (alignment 16) may result in an unaligned pointer value
[-Waddress-of-packed-mem
ber]
  471 |    CONTAINER_OF(ko, struct dyn_obj, kobj);

To avoid the warning, use an intermediate void * variable.

More info in #16587.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

cvinayak commented:
@carlescufi added commits to address required changes related to Broadcast ISO and Connected ISO implementations.

carlescufi commented on kernel/mmu.c:
```diff
@@ -380,7 +380,12 @@ static void free_page_frame_list_put(struct z_page_frame *pf)
 		 "unavailable page put on free list");
-	sys_slist_append(&free_page_frame_list, &pf->node);
+	/* TODO: Check why this assertion hits */
+	/*PF_ASSERT(pf, IS_PTR_ALIGNED(pf, void*), "unaligned pf pointer");*/
```
@dcpleung @andyross any idea why we are hitting this assert if I uncomment it?

andyross replied:
I'm not the expert on paging, but per the comment at the top of the struct in kernel/include/mmu.h it's intentionally misaligned for size optimization.  z_page_frame is a 5- or 9-byte struct (one pointer and a flags byte) and the intent seems to be to pack them all together, alignment be damned.

carlescufi replied:
OK, I will leave out the assertion then. Thanks!

carlescufi commented on kernel/userspace.c:
```diff
@@ -466,8 +466,12 @@ static void unref_check(struct z_object *ko, uintptr_t index)
 
 #ifdef CONFIG_DYNAMIC_OBJECTS
+	/* TODO: check why this assert hits */
+	/*__ASSERT(IS_PTR_ALIGNED(ko, struct dyn_obj), "unaligned z_object");*/
```
@dcpleung @andyross any idea why we are hitting this assert if I uncomment it?

andyross replied:
You want to be checking IS_PTR_ALIGNED(dyn, struct dyn_obj), no?  It would be pathological, but you could imagine a kernel object aligned on N*8 + 4, with a (8-byte-aligned) dyn_obj field offset 4 or 12 bytes in, and that would be legal.

Beyond that possibility, the question is why this particular kernel object got misaligned.  And as this is a dynamic object, that points to a glitch in the allocator?  Someone should definitely look at this.

dcpleung replied:
could be that the object is not allocated using the aligned alloc function?

carlescufi replied:
>You want to be checking IS_PTR_ALIGNED(dyn, struct dyn_obj),

It really is the same, because the `struct z_object` is the first member of `struct dyn_obj`. But sure, let me fix that.

carlescufi replied:
>could be that the object is not allocated using the aligned alloc function?

@dcpleung I have no idea, could you check?

dcpleung replied:
`z_impl_k_object_alloc()` is using `obj_align_get()` to feed alignment to memory allocation, where it should align to `sizeof(void *)`.

Does `IS_PTR_ALIGNED(dyn, struct dyn_obj)` work?

carlescufi commented:
> @carlescufi added commits to address required changes related to Broadcast ISO and Connected ISO implementations.

Thank you @cvinayak, added that in the Bluetooth commit.

cvinayak reviewed (approved):
Controller changes look good.

jhedberg commented on kernel/mmu.c:
```diff
@@ -380,7 +380,12 @@ static void free_page_frame_list_put(struct z_page_frame *pf)
+	/* TODO: Check why this assertion hits */
+	/*PF_ASSERT(pf, IS_PTR_ALIGNED(pf, void*), "unaligned pf pointer");*/
+	/* The structure is packed, which ensures that this is true */
+	void *node = (void *)pf;
```
I don't think the explicit typecast is needed. Isn't one of the main feature of void pointers that you don't need to do explicit casts to/from them.

carlescufi replied:
Absolutely, will check and remove. Thanks!

carlescufi replied:
Should be done now @jhedberg, thanks for catching this.

carlescufi replied:
I've also removed them from the inverse assignment (void to a data type) since those are not necessary either.

jhedberg commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -3728,16 +3728,23 @@ static void le_cis_request(struct pdu_data *pdu_data,
+	/* Check for pdu field being aligned before accessing CIS established
+	 * event.
+	 */
+	node = (void *)pdu_data;
```
Same here regarding unnecessary explicit typecast

jhedberg commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -3769,10 +3777,16 @@ static void le_cis_established(struct pdu_data *pdu_data,
+	/* Check for pdu field being aligned before accessing CIS established
+	 * event.
+	 */
+	node = (void *)pdu_data;
```
And here. I guess it doesn't make sense to to go pointing out every single place in this PR, but if you agree it be good to remove these explicit casts.

carlescufi committed:
x86: acpi: Fix address-of-packed-mem warning

The warning below appears once -Waddress-of-packed-mem is enabled:

/home/carles/src/zephyr/zephyr/arch/x86/core/acpi.c: In function
'z_acpi_find_table':
/home/carles/src/zephyr/zephyr/arch/x86/core/acpi.c:190:24: warning:
taking address of packed member of 'struct acpi_xsdt' may result in an
unaligned pointer value [-Waddress-of-packed-member]
  190 |    for (uint64_t *tp = &xsdt->table_ptrs[0]; tp < end; tp++) {

To avoid the warning, use an intermediate void * variable.

More info in #16587.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
gpio: sx1509b: Fix address-of-packed-mem warning

The warning below appears once -Waddress-of-packed-mem is enabled:

/__w/zephyr/zephyr/drivers/gpio/gpio_sx1509b.c: In function
'port_write':
/__w/zephyr/zephyr/drivers/gpio/gpio_sx1509b.c:456:19: error: taking
address of packed member of 'struct sx1509b_pin_state' may result in an
unaligned pointer value [-Werror=address-of-packed-member]
  456 |  uint16_t *outp = &drv_data->pin_state.data;

To avoid the warning, use an intermediate void * variable.

More info in #16587.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
tests: kernel: Fix address-of-packed-mem warning

The warning below appears once -Waddress-of-packed-mem is enabled:

/__w/zephyr/zephyr/tests/kernel/mem_protect/userspace/src/main.c: In
function 'test_main':
/__w/zephyr/zephyr/tests/kernel/mem_protect/userspace/src/main.c:1024:17:
error: converting a packed 'k_thread_stack_t' {aka 'struct
z_thread_stack_element'} pointer (alignment 1) to a 'struct
z_x86_thread_stack_header' pointer (alignment 4096) may result in an
unaligned pointer value [-Werror=address-of-packed-member]
 1024 |  hdr = ((struct z_x86_thread_stack_header *)ztest_thread_stack);

To avoid the warning, use an intermediate void * variable.

More info in #16587.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
tests: lib: mpsc_pbuf: Fix address-of-packed-mem warning

The warning below appears once -Waddress-of-packed-mem is enabled:

/__w/zephyr/zephyr/tests/lib/mpsc_pbuf/src/main.c: In function
'item_put_data_overwrite':
/__w/zephyr/zephyr/tests/lib/mpsc_pbuf/src/main.c:497:3: error:
converting a packed 'struct test_data_ext' pointer (alignment 1) to a
'uint32_t' {aka 'const unsigned int'} pointer (alignment 4) may result
in an unaligned pointer value [-Werror=address-of-packed-member]
  497 |   mpsc_pbuf_put_data(&buffer, (uint32_t *)&item, len);

To avoid the warning, as well as several others related to the same
problem, use an intermediate void * variable.

More info in #16587.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi committed:
sensor: bmi160: Fix address-of-packed-mem warning

The warning below appears once -Waddress-of-packed-mem is enabled:

/__w/zephyr/zephyr/drivers/sensor/bmi160/bmi160.c: In function
bmi160_gyr_channel_get:
/__w/zephyr/zephyr/drivers/sensor/bmi160/bmi160.c:795:23: error: taking
address of packed member of struct <anonymous> may result in an
unaligned pointer value [-Werror=address-of-packed-member]
  795 |           data->sample.gyr, val);
      |           ~~~~~~~~~~~~^~~~
/__w/zephyr/zephyr/drivers/sensor/bmi160/bmi160.c: In function
bmi160_acc_channel_get:
/__w/zephyr/zephyr/drivers/sensor/bmi160/bmi160.c:807:23: error: taking
address of packed member of struct <anonymous> may result in an
unaligned pointer value [-Werror=address-of-packed-member]
  807 |           data->sample.acc, val);
      |           ~~~~~~~~~~~~^~~~

To avoid the warning, make the struct non-packed, since it is not
necessary in this case given that a union already guarantees that the
pointer to the union points to each member of the union equally..

More info in #16587.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

carlescufi commented on drivers/sensor/bmi160/bmi160.c:
```diff
@@ -790,9 +790,11 @@ static inline void bmi160_gyr_channel_get(const struct device *dev,
-	bmi160_channel_convert(chan, data->scale.gyr,
-			       data->sample.gyr, val);
+	/* Assert below triggers with native_posix. */
+	/*__ASSERT_NO_MSG(IS_PTR_ALIGNED(gyr, uint16_t));*/
```
@andyross @dcpleung @peter-mitsis  Again, I am hitting this on native_posix which is x86-32.

rlubos reviewed (approved):
LGTM

andyross commented on kernel/userspace.c:
```diff
@@ -466,8 +466,11 @@ static void unref_check(struct z_object *ko, uintptr_t index)
+
+	struct dyn_obj *dyn = CONTAINER_OF(vko, struct dyn_obj, kobj);
+	/* TODO: check why this assert hits */
+	/*__ASSERT(IS_PTR_ALIGNED(dyn, struct dyn_obj), "unaligned z_object");*/
```
Just to repeat my comment from elsewhere: this points to some suspicious behavior in the kernel dynamic object allocator.  We should probably open a bug to make sure it gets looked at.  Given the architecture, it's not illegal to misalign structs, but we shouldn't be doing it without a good reason.

(If I had to make a wild guess, it was written for i386 and 4-byte-aligns everything, so misses half the time for structs with 64 bit pointers.)

carlescufi replied:
Indeed, see here: https://github.com/zephyrproject-rtos/zephyr/issues/41062
